src/stcal/ramp_fitting/src/slope_fitter.c

```diff
@@ -0,0 +1,3561 @@
+
+    pr->pixeldq = rd->get_pixeldq(rd->pixeldq, row, col);
+
+    pr->gain = (real_t)rd->get_gain(rd->gain, row, col);
+    if (pr->gain <= 0. || isnan(pr->gain)) {
+        pr->pixeldq |= (rd->dnu | rd->ngval);
+    }
+    for (integ=0; integ<rd->nints; ++integ) {
+        pr->rateints[integ].dq = pr->pixeldq;
+    }
+    pr->rnoise = (real_t) rd->get_rnoise(rd->rnoise, row, col);
+    pr->dcurrent = (real_t) rd->get_dcurrent(rd->dcurrent, row, col);
+    pr->rate.dq = pr->pixeldq;
+}
+
+static void
+get_pixel_ramp_zero(
+        struct pixel_ramp * pr)
+{
+    pr->pixeldq = 0.;
+    pr->gain = 0.;
+    pr->rnoise = 0.;
+
+    /* Zero out flags */
+    memset(pr->is_zframe, 0, pr->nints * sizeof(pr->is_zframe[0]));
+    memset(pr->is_0th , 0, pr->nints * sizeof(pr->is_0th[0]));
+
+    /* C computed values */
+    pr->median_rate = 0.;  /* The median rate of the pixel */
+    pr->invvar_e_sum = 0.; /* Intermediate calculation needed for final slope */
+
+    memset(pr->rateints, 0, pr->nints * sizeof(pr->rateints[0]));
+    memset(&(pr->rate), 0, sizeof(pr->rate));
+}
+
+static void
+get_pixel_ramp_integration_segments_and_pedestal(
+        npy_intp integ,
+        struct pixel_ramp * pr,
+        struct ramp_data * rd)
+{
+    npy_intp idx, idx_pr;
+    real_t fframe, int_slope;
+
+    /* Add list to ramp data structure */
+    idx = get_cube_index(rd, integ, pr->row, pr->col);
+    rd->segs[idx] = pr->segs[integ].head;
+    if (pr->segs[integ].size > rd->max_num_segs) {
+        rd->max_num_segs = pr->segs[integ].size;
+    }
+
+    /* Remove list from pixel ramp data structure */
+    pr->segs[integ].head = NULL;
+    pr->segs[integ].tail = NULL;
+
+    /* Get pedestal */
+    if (pr->rateints[integ].dq & rd->sat) {
+        rd->pedestal[idx] = 0.;
+        return;
+    }
+
+    idx_pr = get_ramp_index(rd, integ, 0);
+    fframe = pr->data[idx_pr];
+    int_slope = pr->rateints[integ].slope;
+
+    // tmp = ((rd->nframes + 1) / 2. + rd->dropframes) / (rd->nframes + rd->groupgap);
+    rd->pedestal[idx] = fframe - int_slope * rd->ped_tmp;
+
+    if (isnan(rd->pedestal[idx])) {
+        rd->pedestal[idx] = 0.;
+    }
+}
+
+/*
+ * This function takes in the args object to parse it, validate the input
+ * arguments, then fill out a ramp data structure to be used for ramp fitting.
+ */
+static struct ramp_data *
+get_ramp_data(
+        PyObject * args)
+{
+    struct ramp_data * rd = calloc(1, sizeof(*rd));  /* Allocate memory */
+    PyObject * Py_ramp_data;
+    char * msg = "Couldn't allocate memory for ramp data structure.";
+
+    /* Make sure memory allocation worked */
+    if (NULL==rd) {
+        PyErr_SetString(PyExc_MemoryError, msg);
+        err_ols_print("%s\n", msg);
+        return NULL;
+    }
+    
+    if (get_ramp_data_parse(&Py_ramp_data, rd, args)) {
+        FREE_RAMP_DATA(rd);
+        return NULL;
+    }
+
+    if (get_ramp_data_arrays(Py_ramp_data, rd)) {
+        FREE_RAMP_DATA(rd);
+        return NULL;
+    }
```

Comment:
`rd` members are populated in both `get_ramp_data_arrays` and `get_ramp_data_parse`. Why not do it all in one function? This division seems arbitrary.

Reply:
The first function parses the arguments passed to the C extension module.  The second function does not parse the C extension arguments, but sets variables from the structure retrieved during argument parsing.